### PR TITLE
Implemented functionality to downgrade to a specific version

### PR DESCRIPTION
**Pull request by carmilso**
The option [-v <version>] allows to downgrade a package to a specific version. If the version is in cache too you must choose which of them you want (local or remote).

e.g.
    downgrade wpa_supplicant -v 2.3-1 (not installed, it downgrades it directly)
    downgrade wpa_supplicant -v 1:2.6-2 (currently installed, it shows two options)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -24,7 +30,7 @@ Downgrade a package, looking in only the A.R.M.:
 $ NOCACHE=1 downgrade foo
 ```
 
-Downgrade a package, looking only in local cache, and favoring `su` over 
+Downgrade a package, looking only in local cache, and favoring `su` over
```
Comment (pbrisbin) [inline README.md]: Thanks for cleaning up the whitespace -- do you might moving those changes to their own commit?

File: downgrade [hunk 2]
```diff
@@ -75,12 +75,14 @@ search_packages() {
   ((NOARM)) || \
     curl --fail --silent "$index" | sed '
       /.* href="\('"$1"'.*\(any\|'"$ARCH"'\)\.pkg\.tar\.xz\)".*/!d;
-      s||'"$index"'\1|g; s|+| |g; s|%|\\x|' | xargs -0 printf "%b"
+      s||'"$index"'\1|g; s|+| |g; s|%|\\x|' | xargs -0 printf "%b" | grep "\-$version"
```
Comment (pbrisbin) [inline downgrade]: I'd use `grep -F` here to avoid things in `$version` being treated like a regex.

File: downgrade [hunk 3]
```diff
@@ -103,8 +113,8 @@ main() {
   for term; do
     candidates=( $(search_packages "$term" | sort_packages) )
 
-    if present_packages "${candidates[@]}"; then
-      if choice="$(read_selection "${candidates[@]}")"; then
+    if check_packages || test "${#candidates[@]}" -gt 0; then
+      if check_selection || choice="${candidates[0]}"; then
```
Comment (pbrisbin) [inline downgrade]: I'm having a little trouble following the intent of these nested, combined conditionals now. Can you explain what this code is doing? Then I might have some suggestions for clarifying the naming.

**Comment by pbrisbin** [comment]
Hey, sorry I didn't see this until now! This is a great addition, and the patch looks good aside from a few comments. Let me know when you've addressed them.

File: downgrade [hunk 4]
```diff
@@ -135,16 +145,18 @@ declare -a terms
 declare -a to_ignore
 declare -a to_install
 declare -a pacman_args
+declare -a version
 
 while [[ -n "$1" ]]; do
   case "$1" in
     --) shift; pacman_args=( "$@" ); break ;;
+    -v) shift; version=( "$1" ); break ;;
```
Comment (pbrisbin) [inline downgrade]: You also shouldn't `break` out of the loop on `-v` or an invocation like `downgrade -v ... -- pacman-options` would never get to the pacman options.

**Comment by pbrisbin** [comment]
Oh and one last thing: can you add a test (or a few) for this feature?

File: README.md (hunk 1 above)
Comment (carmilso) [inline README.md]: Oh! It's done automatically by my vim configuration

File: downgrade (hunk 2 above)
Comment (carmilso) [inline downgrade]: You're right. Great improvement!

File: downgrade (hunk 3 above)
Comment (carmilso) [inline downgrade]: check_packages executes the function present_packages only if there is more than one possibility to download (e.g. if you search a incomplete version like 2.3 being versions 2.3-1 or 2.3-2 since the search of the version is a grep). Then, if there is more than one version (test "${#candidates[@]}" -gt 0) and after being potentially executed present_packages, it continues with the program.
check_selection executes the function read_selection only if there is more than one possibility to download (the same case than above). If there is only one possibility, then the choice is that version and it is not needed to ask the user the number of the version to download (check_selection || choice="${candidates[0]}")

File: downgrade (hunk 4 above)
Comment (carmilso) [inline downgrade]: Yes, in this version I assume that version couldn't be an array but if the pull request is accepted I can work on a version where each version corresponds to the corresponding package in the array. This works now basically to automate the process of downgrading a package when you know the specific version you want.
And you are right again about the break, bug...

**Comment by carmilso** [comment]
How are you implementing the tests for this program? I have tried to downgrade some packages to a specific version and it has worked well (downgrading the package directly to that version)

**Comment by pbrisbin** [comment]
This project uses [cram](https://bitheap.org/cram/) and you can find its current tests as examples in the [test](https://github.com/pbrisbin/downgrade/tree/master/test) directory, and see how they're run via the [CI config](https://github.com/pbrisbin/downgrade/blob/master/circle.yml).

The long and short of it is:

- Write a `foo.t` to hold your test(s)
- Each test is a command to run and expected output
- If the output doesn't match expected, it fails

File: downgrade (hunk 3 above)
Comment (pbrisbin) [inline downgrade]: I see. I think it's confusing to hide the "does thing if there's more than one option" behind these `check_` function. WDYT of writing this something like this:

```bash
if (( "${#candidates[@]" == 1 )); then
  choice=${candidates[0]}
else
  # do what it does now with present/read-selection/etc
  choice=...
end

# proceed, operating on `$choice`
```

It might require a little shuffling of the loop's `continue` and setting of `exit_code`, but I think it'd be worthwhile, and hopefully there are tests to ensure you don't break it.

**Comment by pbrisbin** [comment]
This [blog post](https://pbrisbin.com/posts/mocking_bash/) of mine might be useful, as this is the project that motivated and still uses that strategy.

**Comment by carmilso** [comment]
Excuse me, but I have not enough time now to realize all this modifications. I hope fixing it in few weeks

**Comment by pbrisbin** [comment]
@carmilso I began a new branch based off yours, intending to just make those fixups, but I ended up taking things in a slightly different direction to accomplish this feature. Please let me know what you think: #52

**Comment by carmilso** [comment]
I have revised it and I agree with all of that modifications from my proposal. I will try to collaborate with this improvement!

**Comment by pbrisbin** [comment]
This has been implemented via #52 and released as 5.4.0. Thanks for the inspiration!
